core/src/bitcoin_syncer.rs

```diff
@@ -0,0 +1,205 @@
+    let block_hash = rpc.client.get_block_hash(height).await?;
+    let block_header = rpc.client.get_block_header(&block_hash).await?;
+    Ok(BlockInfo {
+        block_hash,
+        block_header,
+        block_height: height,
+    })
+}
+
+pub async fn process_block(
+    db: &Database,
+    block: &bitcoin::Block,
+    block_height: i64,
+) -> Result<(), BridgeError> {
+    let mut dbtx = db.begin_transaction().await?;
+    let block_hash = block.header.block_hash();
+
+    db.set_chain_head(
+        Some(&mut dbtx),
+        &block_hash,
+        &block.header.prev_blockhash,
+        block_height,
+    )
+    .await?;
+
+    for tx in &block.txdata {
+        process_tx(db, &mut dbtx, tx, &block_hash).await?;
+    }
+
+    dbtx.commit().await?;
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn process_tx(
+    db: &Database,
+    dbtx: DatabaseTransaction<'_, '_>,
+    tx: &bitcoin::Transaction,
+    block_hash: &bitcoin::BlockHash,
+) -> Result<(), BridgeError> {
+    let txid = tx.compute_txid();
+    db.insert_tx(dbtx, block_hash, &txid).await?;
+
+    for input in &tx.input {
+        db.insert_utxo(
+            dbtx,
+            &txid,
+            &input.previous_output.txid,
+            input.previous_output.vout as i64,
+        )
+        .await?;
+    }
+
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn start_bitcoin_syncer(
+    db: Database,
+    rpc: ExtendedRpc,
+    poll_delay: Duration,
+    mode: BitcoinSyncerPollingMode,
+) -> Result<ChainHeadPollingResult, BridgeError> {
+    let (tx, _) = broadcast::channel(100);
+
+    let returned_tx = tx.clone();
+
+    let mut block_height = db
+        .get_max_height(None)
+        .await?
+        .ok_or(BridgeError::BlockNotFound)?;
+
+    let handle = tokio::spawn(async move {
+        loop {
+            let mut block_hash = rpc.client.get_block_hash(block_height + 1).await?;
+            let mut block_header = rpc.client.get_block_header(&block_hash).await?;
+
+            let mut new_blocks = vec![BlockInfo {
+                block_hash,
+                block_header,
+                block_height: block_height + 1,
+            }];
+
+            for _ in 0..100 {
+                // if the previous block is in the db, do nothing
+                let height = db
+                    .get_height_from_block_hash(None, block_header.prev_blockhash)
+                    .await?;
+                if height.is_some() {
+                    break;
+                }
+
+                // if the previous block is not in the db, we need to get the previous block
+
+                block_hash = block_header.prev_blockhash;
+                block_header = rpc.client.get_block_header(&block_hash).await?;
+
+                let block_info = BlockInfo {
+                    block_hash,
+                    block_header,
+                    block_height,
+                };
+                new_blocks.push(block_info);
+
+                block_height -= 1;
+            }
```

Comment:
this logic is too complicated. we need to simplify it. It also depends on a global variable (block_height).

I'm also not too sure on its correctness here. `block_height` starts off as the longest chain in our db. Then for each block that you can't find, you'll subtract the height by one going back on the longest chain. I feel like the `block_height` ought to start from block_height + 1 (which is the height of the first block_header that we fetch).

Overall, this logic is way too complicated and uses global variables which really complicates things.

I believe this can be extracted to something like:

```rust
async fn get_next_block(db) -> (Vec<Blocks> /* reorg */, Option<Block> /* one new block */)
```

Reply:
Actually, I guess the second item would be a `Vec<Block>` of new blocks. It might be good to have an enum return:

```rust
enum BlockPollResult {
  OneNewBlock(BlockHeader),
  ReorgedBlocks(Vec<Block>, Vec<Block>)
}
```

Reply:
also better naming for function:
```rs
async fn poll_blocks()
```